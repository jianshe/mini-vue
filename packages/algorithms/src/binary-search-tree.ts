export default class Stack {
  private _count: number;
  private _items: {};

  constructor() {
    this._count = 0;
    this._items = {};
  }

  push(element: any) {
    this._items[this._count] = element;
    this._count++;
  }
  pop() {
    delete this._items[this._count];
    this._count--;
  }
  peek() {
    return this._items[this._count - 1];
  }
  size() {
    return this._count;
  }
  isEmpty() {
    return this._count === 0;
  }
  clear() {
    this._items = {};
    this._count = 0;
  }
  toString() {
    let toStringStr = this._items[0];
    for (let i = 1; i < this._count; i++) {
      toStringStr = `${toStringStr},${this._items[i]}`;
    }
    return toStringStr
  }
}
