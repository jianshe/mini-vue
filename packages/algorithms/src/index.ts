export { countingSort } from "./counting-sort";
export { quickSort } from "./quick-sort";
