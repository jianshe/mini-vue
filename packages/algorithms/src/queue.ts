export default class Queue<T> {
  private _count: number;
  private _mixPopNumber: number;
  private _items: {};
  // FIFO 先进先出
  constructor() {
    this._count = 0;
    this._mixPopNumber = 0;
    this._items = {};
  }
  // 入队
  enQueue(element) {
    this._items[this._count] = element;
    this._count++;
  }
  deQueue() {
    if (this.isEmpty()) {
      return undefined;
    }
    this._count--;
    const result = this._items[this._mixPopNumber];
    delete this._items[this._mixPopNumber];
    this._mixPopNumber++;
    return result;
  } //出队
  isEmpty() {
    return this._count === 0;
  }
  clear() {
    this._count = 0;
    this._items = {};
    this._mixPopNumber = 0;
  }
  size() {
    return this._count - this._mixPopNumber;
  }
  peek() {
    if (this.isEmpty()) {
      return undefined;
    }
    return this._items[this._mixPopNumber];
  } // 返回队列中的第一个元素--最先被添加，也将是最先被移除的元素
  toObjString() {
    if (this.isEmpty()) {
      return "";
    }
    let objString = `${this._items[this._mixPopNumber]}`;
    for (let i = this._mixPopNumber + 1; i < this._count; i++) {
      objString = `${objString},${this._items[i]}`;
    }
    return objString;
  }
}
