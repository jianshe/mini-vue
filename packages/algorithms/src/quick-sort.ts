import { Compare, defaultCompare, swap } from "./util";

function quick(array, left, right, comparedFn = defaultCompare) {
  let index;
  if (array.length > 1) {
    index = partial(array, left, right, comparedFn);
    if (left < index - 1) {
      quick(array, left, index - 1, comparedFn);
    }
    if (right > index) {
      quick(array, index, right, comparedFn);
    }
  }
  return array;
}

function partial(array, left, right, compareFn) {
  let pivot = array[Math.floor((left + right) / 2)];
  let i = left;
  let j = right;
  while (i <= j) {
    while (compareFn(array[i], pivot) === Compare.LESS_THAN) {
      i++;
    }
    while (compareFn(array[j], pivot) === Compare.BIGGER_THAN) {
      j--;
    }
    if (i <= j) {
      swap(array, i, j);
      i++;
      j--;
    }
  }
  return i;
}

export function quickSort(array: any, comparedFn = defaultCompare) {
  return quick(array, 0, array.length - 1, comparedFn);
}
