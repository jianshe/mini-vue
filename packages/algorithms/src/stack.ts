// stack FILO 先进后出
export default class Stack {
  private items: any = [];
  constructor() {
    this.items = [];
  }

  push(element: any) {
    this.items.push(element);
  }

  pop() {
    return this.items.pop();
  }
  peek() {
    // 获取栈顶元素
    return this.items.unshift();
  }
  isEmpty() {
    return this.items.length === 0;
  }

  toString() {
    return this.items.join(",");
  }

  clear() {
    this.items = [];
  }

  size() {
    return this.items.length;
  }
}
