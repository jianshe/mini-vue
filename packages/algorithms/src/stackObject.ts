export default class Stack {
  private _count: number;
  private _items: {};
  constructor() {
    this._count = 0;
    this._items = {};
  }
  push(element) {
    this._items[this._count] = element;
    this._count++;
  }
  pop() {
    if (this.isEmpty()) {
      return undefined;
    }
    this._count--;
    const result = this._items[this._count];
    delete this._items[this._count];
    return result;
  }
  peek() {
    if (this.isEmpty()) {
      return undefined;
    }
    return this._items[this._count - 1];
  }
  clear() {
    this._items = {};
    this._count = 0;
  }
  isEmpty() {
    return this._count === 0;
  }
  size() {
    return this._count;
  }
  toString() {
    if (this.isEmpty()) {
      return "";
    }
    let objString = this._items[0];
    for (let i = 1; i < this._count; i++) {
      objString = `${objString},${this._items[i]}`;
    }
    return objString;
  }
}
