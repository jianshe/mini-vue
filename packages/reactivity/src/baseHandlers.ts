import { extend, isObject } from "@guide-mini-vue/shared";
import { track, trigger } from "./effect";
import { ReactiveFlags, reactive, readonly } from "./reactive";
const get = createGetter();
const set = createSetter();
const readonlyGet = createGetter(true);
const shallowReadonlyGet = createGetter(true, true);
function createGetter(isReadonly = false, shallow = false) {
  return function get(target, key) {
    const res = Reflect.get(target, key);
    if (key === ReactiveFlags.IS_REACTIVE) {
      return !isReadonly;
    } else if (key === ReactiveFlags.IS_READONLY) {
      return isReadonly;
    }
    if (!isReadonly) {
      // 在触发 get 的时候进行依赖收集
      track(target, key);
    }

    if (shallow) {
      return res;
    }

    if (isObject(res)) {
      // 把内部所有的是 object 的值都用 reactive 包裹，变成响应式对象
      // 如果说这个 res 值是一个对象的话，那么我们需要把获取到的 res 也转换成 reactive
      // res 等于 target[key]
      return isReadonly ? readonly(res) : reactive(res);
    }
    return res;
  };
}

function createSetter() {
  return function set(target, key, value) {
    const result = Reflect.set(target, key, value);

    trigger(target, key);

    return result;
  };
}

export const readonlyHandlers = {
  get: readonlyGet,
  set(target, key) {
    // readonly 的响应式对象不可以修改值
    console.warn(
      `Set operation on key "${String(key)}" failed: target is readonly.`,
      target
    );
    return true;
  },
};

export const mutableHandlers = {
  get,
  set,
};

export const shallowReadonlyHandlers = extend(
  {},
  readonlyHandlers,
  shallowReadonlyGet
);
