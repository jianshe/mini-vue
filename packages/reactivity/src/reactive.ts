import { isObject } from "@guide-mini-vue/shared";
import { mutableHandlers, readonlyHandlers, shallowReadonlyHandlers } from "./baseHandlers";

export const enum ReactiveFlags {
  IS_REACTIVE = "__v_isReactive",
  IS_READONLY = "__v_isReadonly",
}

export function reactive(raw) {
  return createReactiveObject(raw, mutableHandlers);
}

export function readonly(raw) {
  return createReactiveObject(raw, readonlyHandlers);
}
export function shallowReadonly(raw) {
  return createReactiveObject(raw, shallowReadonlyHandlers);
}

export function isReadonly(value) {
  return !!value[ReactiveFlags.IS_READONLY];
}

export function isReactive(value) {
  // 如果 value 是 proxy 的话
  // 会触发 get 操作，而在 createGetter 里面会判断
  // 如果 value 是普通对象的话
  // 那么会返回 undefined ，那么就需要转换成布尔值
  return !!value[ReactiveFlags.IS_REACTIVE];
}

export function isProxy(value) {
  return isReactive(value) || isReadonly(value)
}

function createReactiveObject(target, baseHandlers) {
  if (!isObject(target)) {
    console.warn(`target ${target} 必须是一个对象`)
    return
  }
  return new Proxy(target, baseHandlers);
}
