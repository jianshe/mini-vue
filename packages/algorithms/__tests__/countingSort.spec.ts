import { countingSort } from "../src/counting-sort";
import { createNonSortedArray } from "../src/util";

describe("counting sort", () => {
  it("happy path", () => {
    const nonSortArray: any[] = createNonSortedArray(9);

    const sortedArray = countingSort(nonSortArray);
    console.log("sortedArray", sortedArray);
    expect(sortedArray).toMatchSnapshot();
  });
});
