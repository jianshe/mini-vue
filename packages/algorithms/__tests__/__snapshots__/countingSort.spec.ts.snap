// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`counting sort > happy path 1`] = `
[
  6,
  8,
  9,
  10,
  22,
  23,
  98,
]
`;
